core/src/util/shape.rs

```diff
@@ -0,0 +1,243 @@
+}
+
+impl AsShape for syn::FieldsUnnamed {
+    fn as_shape(&self) -> Shape {
+        if self.unnamed.len() == 1 {
+            Shape::Newtype
+        } else {
+            Shape::Tuple
+        }
+    }
+}
+
+impl AsShape for syn::DataStruct {
+    fn as_shape(&self) -> Shape {
+        self.fields.as_shape()
+    }
+}
+
+impl AsShape for syn::Variant {
+    fn as_shape(&self) -> Shape {
+        self.fields.as_shape()
+    }
+}
+
+/// Description of how fields in a struct or variant are syntactically laid out.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Shape {
+    /// A set of named fields, e.g. `{ field: String }`.
+    Named,
+    /// A list of unnamed fields, e.g. `(String, u64)`.
+    Tuple,
+    /// No fields, e.g. `struct Example;`
+    Unit,
+    /// A special case of [`Tuple`](Shape#variant.Tuple) with exactly one field, e.g. `(String)`.
+    Newtype,
+}
+
+impl Shape {
+    pub fn description(&self) -> &'static str {
+        match self {
+            Shape::Named => "named fields",
+            Shape::Tuple => "unnamed fields",
+            Shape::Unit => "no fields",
+            Shape::Newtype => "one unnamed field",
+        }
+    }
+}
+
+impl AsShape for Shape {
+    fn as_shape(&self) -> Shape {
+        *self
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ShapeSet {
+    newtype: bool,
+    named: bool,
+    tuple: bool,
+    unit: bool,
+    any: bool,
+}
+
+impl ShapeSet {
+    pub fn set_any(&mut self, any: bool) -> &mut Self {
+        self.any = any;
+        self
+    }
+
+    pub fn insert(&mut self, shape: Shape) {
+        match shape {
+            Shape::Named => self.named = true,
+            Shape::Tuple => self.tuple = true,
+            Shape::Unit => self.unit = true,
+            Shape::Newtype => self.newtype = true,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        !self.any && !self.named && !self.newtype && !self.tuple && !self.unit
+    }
+
+    fn matches_shape(&self, shape: Shape) -> bool {
+        if self.any {
+            return true;
+        }
+
+        match shape {
+            Shape::Named => self.named,
+            Shape::Tuple => self.tuple,
+            Shape::Unit => self.unit,
+            Shape::Newtype => self.newtype || self.tuple,
+        }
+    }
+
+    pub fn matches(&self, fields: &impl AsShape) -> bool {
+        self.matches_shape(fields.as_shape())
+    }
+
+    /// Check if a container matches this set of allowed shapes, and produce
+    /// an [`Error`](crate::Error) if it does not.
+    pub fn check(&self, fields: &impl AsShape) -> crate::Result<()> {
+        let shape = fields.as_shape();
+
+        if self.matches_shape(shape) {
+            Ok(())
+        } else {
+            Err(crate::Error::unsupported_shape_with_expected(
+                shape.description(),
+                self,
+            ))
+        }
+    }
+}
+
+impl fmt::Display for ShapeSet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
It would be nice to build this as a more fluent sentence, e.g. "named fields or unnamed fields" in the case of two items, or "named fields, unnamed fields, or no fields" when three options are applicable.

Reply:
Done